File: pwiz/analysis/demux/PrecursorMaskCodec.cpp
//
// $Id$
//
//
// Original author: Jarrett Egertson <jegertso .@. uw.edu>
//
// Licensed under the Apache License, Version 2.0 (the "License"); 
// you may not use this file except in compliance with the License. 
// You may obtain a copy of the License at 
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software 
// distributed under the License is distributed on an "AS IS" BASIS, 
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
// See the License for the specific language governing permissions and 
// limitations under the License.
//

#include "PrecursorMaskCodec.hpp"
#include "DemuxDataProcessingStrings.hpp"

namespace pwiz{
namespace analysis{
    using namespace std;
    using namespace msdata;
    using namespace Eigen;

    PrecursorMaskCodec::PrecursorMaskCodec(SpectrumList_const_ptr slPtr, Params p)
        : spectraPerCycle_(0),
        precursorsPerSpectrum_(0),
        overlapsPerSpectrum_(0),
        params_(p)
    {
        ReadDemuxScheme(slPtr);
    }

    template <typename T>
    void PrecursorMaskCodec::GetMask(T&& arrayType, Spectrum_const_ptr sPtr, double weight) const
    {
        vector<size_t> indices;
        SpectrumToIndices(sPtr, indices);
        if (params_.variableFill)
        {
            vector<DemuxWindow> demuxWindows;
            for (auto i : indices)
            {
                demuxWindows.push_back(isolationWindows_[i].demuxWindow); // cache the demux windows
            }
            for (const auto& p : sPtr->precursors)
            {
                DemuxWindow precursorWindow(p);
                for (auto i = 0; i < indices.size(); ++i)
                {
                    if (precursorWindow.ContainsCenter(demuxWindows[i]))
                    {
                        arrayType[indices[i]] = weight * p.userParam("MultiFillTime").valueAs<double>() / 1000.0;
                        break;
                    }
                }
            }
        }
        else
        {
            for (auto index : indices)
            {
                arrayType[index] = weight;
            }
        }
    }

    VectorXd PrecursorMaskCodec::GetMask(msdata::Spectrum_const_ptr sPtr, double weight) const
    {
        VectorXd maskVector = VectorXd::Zero(GetDemuxBlockSize());
        GetMask(maskVector.data(), sPtr, weight);
        return maskVector;
    }

    void PrecursorMaskCodec::GetMask(msdata::Spectrum_const_ptr sPtr, DemuxTypes::MatrixType& m, size_t rowNum, double weight) const
    {
        m.row(rowNum).setZero();
        GetMask(m.row(rowNum), sPtr, weight);
    }

    void PrecursorMaskCodec::ReadDemuxScheme(msdata::SpectrumList_const_ptr spectrumList)
    {
        IdentifyCycle(spectrumList, isolationWindows_);
        IdentifyOverlap(isolationWindows_);
        // TODO once unique isolation regions have been identified it is possible to recognize patterns in their layout to optimize for speed
    }

    void PrecursorMaskCodec::IdentifyCycle(msdata::SpectrumList_const_ptr spectrumList, vector<IsolationWindow>& demuxWindows)
    {
        Spectrum_const_ptr spec;
        map<string, Precursor> precursorMap;
        string ms2SpectrumMissingPrecursorInfoError("IdentifyCycle() MS2 spectrum is missing precursor information.");
        precursorsPerSpectrum_ = 0;
        {
            size_t index = 0;
            {
                // Find the first MS2 spectrum to use as a representative spectrum
                bool foundAtLeastOneMS2 = false;
                
            for (; index < spectrumList->size(); ++index)
            {
                spec = spectrumList->spectrum(index);
                if (spec->cvParam(MS_ms_level).valueAs<int>() == 2)
                {
                    // Found the first MS2 spectrum, record any relevant qualities
                        if (spec->precursors.size() == 0)
                            throw runtime_error(ms2SpectrumMissingPrecursorInfoError);
                    precursorsPerSpectrum_ = spec->precursors.size();
                        foundAtLeastOneMS2 = true;
                    break;
                }
            }

                if (!foundAtLeastOneMS2)
                throw runtime_error("IdentifyCycle() No MS2 scans found for this experiment.");
            }

            // Continue searching and identifying precursors until all unique precursors are found
            size_t mappedAlready = 0;
            for (; index < spectrumList->size() && mappedAlready <= 2 * precursorMap.size(); index++)
            {
                spec = spectrumList->spectrum(index);
                if (spec->cvParam(MS_ms_level).valueAs<int>() != 2) continue;
                if (spec->precursors.size() == 0)
                    throw runtime_error(ms2SpectrumMissingPrecursorInfoError);
                if (spec->precursors.size() != precursorsPerSpectrum_)
                    throw runtime_error("IdentifyCycle() Number of precursors is varying between individual MS2 scans. Cannot infer demultiplexing scheme.");
                for (const auto& p : spec->precursors)
                {
                    string mzString = prec_to_string(p);
                    auto it = precursorMap.find(mzString);
                    if (it == precursorMap.end())
                    {
                        // precursor window was not already present, add it
                        mappedAlready = 0;
                        precursorMap[mzString] = p;
                    }
                    else
                    {
                        // precursor window was already seen, move on
                        mappedAlready += 1;
                    }
                }
            }
            if (mappedAlready <= 2 * precursorMap.size())
            {
                throw runtime_error("IdentifyCycle() Could not determine demultiplexing scheme. Too few spectra to determine the number of precursor windows.");
            }
        }

        //@{ Get the sorted keys from the map
        vector< string > sortedKeys;
        for (const auto& pe : precursorMap)
        {
            sortedKeys.push_back(pe.first);
        }
        sort(sortedKeys.begin(), sortedKeys.end(), stringToFloatCompare);

        demuxWindows.clear();
        demuxWindows.reserve(sortedKeys.size());
        for (const auto& keyString : sortedKeys)
        {
            demuxWindows.push_back(IsolationWindow(precursorMap[keyString]));
        }
        // precursors is now filled and sorted
        //@}

        if (precursorsPerSpectrum_ == 0)
            throw logic_error("IdentifyCycle() Number of precursors per spectrum is 0.");

        // We can now solve for spectraPerCycle regardless of the presence of overlap
        spectraPerCycle_ = demuxWindows.size() / precursorsPerSpectrum_;

        if (spectraPerCycle_ == 0)
            throw logic_error("IdentifyCycle() Number of spectra per cycle is 0.");
    }

    void PrecursorMaskCodec::IdentifyOverlap(vector<IsolationWindow>& isolationWindows)
    {
        if (isolationWindows.size() <= 1)
            return;

        const MZHash minimumWindowSize = IsoWindowHasher::Hash(params_.minimumWindowSize);

        // Reduce risk of unintentional modification
        const auto& const_isolationWindows = isolationWindows;

        // Record all possible demux window boundaries
        set<DemuxBoundary> demuxBoundaries;
        for (auto it = const_isolationWindows.begin(); it != const_isolationWindows.end(); ++it)
        {
            demuxBoundaries.insert(DemuxBoundary(it->lowMz));
            demuxBoundaries.insert(DemuxBoundary(it->highMz));
        }
        const auto& const_demuxBoundaries = demuxBoundaries;

        // Merge nearby boundaries
        vector<DemuxBoundary> exactBoundaries;
        {
            auto lowMZ = const_demuxBoundaries.begin();
            auto highMZ = const_demuxBoundaries.begin();
            ++highMZ;
            for (; highMZ != const_demuxBoundaries.end(); ++highMZ, ++lowMZ)
            {
                if (highMZ->mzHash - lowMZ->mzHash > minimumWindowSize)
                    exactBoundaries.push_back(DemuxBoundary(lowMZ->mz));
                else
                {
                    // Since this is a small window it is most likely an edge we'll want to use, though not
                    // necessarily. It could be that two windows are adjacent with no common overlaping window.
                    // Though this isn't a likely use case we want to account for it. Later we'll match these
                    // potential edges to centers of real demux windows at which point these edge cases will
                    // be thrown out.
                    double averageMz = (highMZ->mz + lowMZ->mz) / 2.0;
                    exactBoundaries.push_back(DemuxBoundary(averageMz));
                    ++highMZ;
                    ++lowMZ;
                }
            }

            // Add the final boundary
            exactBoundaries.push_back(DemuxBoundary(lowMZ->mz));
        }

        // Generate a set of possible demux windows from the set of boundaries
        vector<IsolationWindow> possibleWindows;
        possibleWindows.reserve(exactBoundaries.size() - 1);
        {
            auto lowMZ = exactBoundaries.begin();
            auto highMZ = exactBoundaries.begin();
            ++highMZ;
            for (; highMZ != exactBoundaries.end(); ++highMZ, ++lowMZ)
            {
                possibleWindows.push_back(IsolationWindow(lowMZ->mz, highMZ->mz));
            }
        }

        // Identify the demux windows contained within each precursor isolation window and the number of times they are used.
        // Record which ranges are used to later track the unused
        multiset<IsolationWindow> usedWindows;
        for (auto it = const_isolationWindows.begin(); it != const_isolationWindows.end(); ++it)
        {
            for (auto subWindowIt = possibleWindows.begin(); subWindowIt != possibleWindows.end(); ++subWindowIt)
            {
                // TODO take advantage of sorting to make this O(n) instead of O(n^2) by not repeating lower and higher ranges
                if (it->demuxWindow.ContainsCenter(subWindowIt->demuxWindow))
                    usedWindows.insert(*subWindowIt);
            }
        }

        // Find the number of overlaps by getting the count for each demux window
        size_t maxCount = 0;
        // Record the windows that were used
        vector<IsolationWindow> returnIsolationWindows;
        for (auto it = usedWindows.begin(); it != usedWindows.end(); )
        {
            auto count = usedWindows.count(*it);
            maxCount = max(maxCount, count);
            returnIsolationWindows.push_back(*it);
            std::advance(it, count);
        }
        overlapsPerSpectrum_ = maxCount;

        // Find the unused windows
        /*
        set<DemuxWindow> unusedWindows;
        set_difference(possibleWindows.begin(), possibleWindows.end(), usedWindows.begin(), usedWindows.end(), inserter(unusedWindows, unusedWindows.begin()));
        */

        if (overlapsPerSpectrum_ == 0)
        {
            throw logic_error("IdentifyOverlap() Number of demux windows is 0.");
        }

        isolationWindows = move(returnIsolationWindows);
    }

    void PrecursorMaskCodec::SpectrumToIndices(msdata::Spectrum_const_ptr spectrumPtr, std::vector<size_t>& indices) const
    {
        if (spectrumPtr->precursors.size() != precursorsPerSpectrum_)
        {
            throw runtime_error("SpectrumToIndices() Number of precursors in this spectrum differ from the number expected for this demultiplexing scheme.");
        }

        indices.clear();
        vector<DemuxWindow> overlappingWindows;
        for (const auto& precursor : spectrumPtr->precursors)
        {
            overlappingWindows.push_back(DemuxWindow(precursor));
        }

        sort(overlappingWindows.begin(), overlappingWindows.end());

        auto searchLowerBoundIt = isolationWindows_.begin();
        for (const auto& window : overlappingWindows)
        {
            for (auto searchIt = searchLowerBoundIt; searchIt != isolationWindows_.end(); ++searchIt)
            {
                if (window.mzHigh <= searchIt->demuxWindow.mzLow)
                {
                    // the search window has passed the multiplexed window entirely. All remaining windows will also be past the multiplexed window, so we
                    // don't need to search them.
                    break;
                }
                if (window.ContainsCenter(searchIt->demuxWindow))
                {
                    indices.push_back(distance(isolationWindows_.begin(), searchIt)); // found a window, add its index to the list
                    searchLowerBoundIt = searchIt + 1; // update our lower bound before we keep searching so that we don't repeat this region
                }
            }
        }
        assert(indices.size() > 0);
        if (indices.size() != overlapsPerSpectrum_ * precursorsPerSpectrum_)
            /* This can happen when either (1) the experimental scheme is not solvable as a demultiplexing problem, or (2) the window
             * boundary tolerance (hardcoded in the IsoWindowHasher) is set too low. Most likely, the users isolation scheme was not
             * properly defined causing window boundaries to not align. For example, using too small or large of an isolation window
             * width for a set of isolation targets can cause misalignment. This can be compensated for to some extent by increasing the
             * minimum window size. This workaround isn't ideal and can create artifacts, but can salvage experiments in a pinch. */
            throw runtime_error("SpectrumToIndices() Number of demultiplexing windows changed. Minimum window size or window boundary tolerance may be set too low.");
    }

    struct IsolationWindow PrecursorMaskCodec::GetIsolationWindow(size_t i) const
    {
        return isolationWindows_[i];
    }

    size_t PrecursorMaskCodec::GetNumDemuxWindows() const
    {
        return isolationWindows_.size();
    }

    int PrecursorMaskCodec::GetSpectraPerCycle() const
    {
        return static_cast<int>(spectraPerCycle_);
    }

    int PrecursorMaskCodec::GetPrecursorsPerSpectrum() const
    {
        return static_cast<int>(precursorsPerSpectrum_);
    }

    int PrecursorMaskCodec::GetOverlapsPerCycle() const
    {
        return static_cast<int>(overlapsPerSpectrum_);
    }

    size_t PrecursorMaskCodec::GetDemuxBlockSize() const
    {
        return spectraPerCycle_ * precursorsPerSpectrum_ * overlapsPerSpectrum_;
    }
} // namespace analysis
} // namespace pwiz

File: pwiz-sharp/src/Analysis/Demux/PrecursorMaskCodec.cs
using MathNet.Numerics.LinearAlgebra;
using MathNet.Numerics.LinearAlgebra.Double;
using Pwiz.Data.Common.Cv;
using Pwiz.Data.MsData.Spectra;

namespace Pwiz.Analysis.Demux;

/// <summary>
/// Hashes a floating-point m/z value to a 64-bit integer with ±5e-9 effective precision.
/// Port of cpp's <c>IsoWindowHasher</c> (<c>IPrecursorMaskCodec.hpp</c>).
/// </summary>
public static class IsoWindowHasher
{
    private const double Multiplier = 1e8;

    /// <summary>Hashes <paramref name="mz"/> as <c>round(mz * 1e8)</c>. m/z 500.49 → 50049000000.</summary>
    public static long Hash(double mz) => (long)System.Math.Round(mz * Multiplier);

    /// <summary>Inverse of <see cref="Hash"/>.</summary>
    public static double UnHash(long hashed) => hashed / Multiplier;
}

/// <summary>
/// Slim m/z-range container used as a key in demux algorithms. Port of cpp's <c>DemuxWindow</c>.
/// Boundaries are stored as <see cref="IsoWindowHasher"/> hashes so equality and ordering are
/// fuzzy-tolerant (±5e-9 m/z).
/// </summary>
public readonly record struct DemuxWindow(long MzLow, long MzHigh) : IComparable<DemuxWindow>
{
    /// <summary>Constructs a <see cref="DemuxWindow"/> from a precursor's isolation window.</summary>
    public DemuxWindow(Precursor p)
        : this(IsoWindowHasher.Hash(DemuxHelpers.PrecursorTarget(p) - DemuxHelpers.PrecursorLowerOffset(p)),
               IsoWindowHasher.Hash(DemuxHelpers.PrecursorTarget(p) + DemuxHelpers.PrecursorUpperOffset(p))) { }

    /// <summary>True iff <paramref name="inner"/>'s m/z range is fully contained in this one.</summary>
    public bool Contains(DemuxWindow inner) => inner.MzLow >= MzLow && inner.MzHigh <= MzHigh;

    /// <summary>True iff <paramref name="inner"/>'s center m/z lies within this window.</summary>
    public bool ContainsCenter(DemuxWindow inner)
    {
        long center = (long)System.Math.Round(inner.MzLow + (inner.MzHigh - inner.MzLow) / 2.0);
        return center >= MzLow && center <= MzHigh;
    }

    /// <inheritdoc/>
    public int CompareTo(DemuxWindow other) => MzLow.CompareTo(other.MzLow);

#pragma warning disable CS1591 // comparison operators delegate to CompareTo; docs would just restate the operator
    public static bool operator <(DemuxWindow a, DemuxWindow b) => a.CompareTo(b) < 0;
    public static bool operator >(DemuxWindow a, DemuxWindow b) => a.CompareTo(b) > 0;
    public static bool operator <=(DemuxWindow a, DemuxWindow b) => a.CompareTo(b) <= 0;
    public static bool operator >=(DemuxWindow a, DemuxWindow b) => a.CompareTo(b) >= 0;
#pragma warning restore CS1591
}

/// <summary>
/// Full-precision wrapper around a <see cref="DemuxWindow"/> that retains the original
/// floating-point m/z bounds. Port of cpp's <c>IsolationWindow</c> in
/// <c>IPrecursorMaskCodec.hpp</c>. Named with the <c>Demux</c> prefix to disambiguate from
/// <c>Pwiz.Data.MsData.Spectra.IsolationWindow</c>.
/// </summary>
public readonly record struct DemuxIsolationWindow(double LowMz, double HighMz, DemuxWindow Window)
    : IComparable<DemuxIsolationWindow>
{
    /// <summary>Constructs from raw m/z bounds.</summary>
    public DemuxIsolationWindow(double lowMz, double highMz)
        : this(lowMz, highMz, new DemuxWindow(IsoWindowHasher.Hash(lowMz), IsoWindowHasher.Hash(highMz))) { }

    /// <summary>Constructs from a precursor's isolation window CV params.</summary>
    public static DemuxIsolationWindow FromPrecursor(Precursor p) =>
        new(DemuxHelpers.PrecursorMzLow(p), DemuxHelpers.PrecursorMzHigh(p), new DemuxWindow(p));

    /// <inheritdoc/>
    public int CompareTo(DemuxIsolationWindow other) => Window.CompareTo(other.Window);

#pragma warning disable CS1591 // comparison operators delegate to CompareTo; docs would just restate the operator
    public static bool operator <(DemuxIsolationWindow a, DemuxIsolationWindow b) => a.CompareTo(b) < 0;
    public static bool operator >(DemuxIsolationWindow a, DemuxIsolationWindow b) => a.CompareTo(b) > 0;
    public static bool operator <=(DemuxIsolationWindow a, DemuxIsolationWindow b) => a.CompareTo(b) <= 0;
    public static bool operator >=(DemuxIsolationWindow a, DemuxIsolationWindow b) => a.CompareTo(b) >= 0;
#pragma warning restore CS1591
}

/// <summary>
/// Generates and accesses precursor masks for a demultiplexing scheme. Port of cpp's
/// <c>IPrecursorMaskCodec</c>.
/// </summary>
public interface IPrecursorMaskCodec
{
    /// <summary>Returns a design-matrix row describing which precursor isolation windows are
    /// present in <paramref name="spectrum"/>, scaled by <paramref name="weight"/>.</summary>
    Vector<double> GetMask(Spectrum spectrum, double weight = 1.0);

    /// <summary>Same as <see cref="GetMask(Spectrum,double)"/> but writes into row
    /// <paramref name="rowNum"/> of the user-provided matrix (avoids a per-call allocation).</summary>
    void GetMask(Spectrum spectrum, Matrix<double> m, int rowNum, double weight = 1.0);

    /// <summary>Maps the precursor isolation windows of <paramref name="spectrum"/> to indices
    /// into the demux design matrix.</summary>
    void SpectrumToIndices(Spectrum spectrum, List<int> indices);

    /// <summary>The isolation window at design-matrix column <paramref name="i"/>.</summary>
    DemuxIsolationWindow GetIsolationWindow(int i);

    /// <summary>Total number of demux'd precursor windows (= number of design-matrix columns).</summary>
    int NumDemuxWindows { get; }

    /// <summary>Indices of edge windows removed during overlap analysis (<see cref="PrecursorMaskCodec.Params.RemoveNonOverlappingEdges"/>).</summary>
    IReadOnlyList<int> DemuxWindowEdgesRemoved { get; }

    /// <summary>Number of spectra required to cover all precursor isolation windows.</summary>
    int SpectraPerCycle { get; }

    /// <summary>Number of precursors per multiplexed spectrum (constant across the run).</summary>
    int PrecursorsPerSpectrum { get; }

    /// <summary>Number of overlap repeats per cycle (1 for non-overlapped, 2 for split-by-half, etc.).</summary>
    int OverlapsPerCycle { get; }

    /// <summary>Total number of windows the design matrix needs to cover. Equals
    /// <see cref="SpectraPerCycle"/> × <see cref="PrecursorsPerSpectrum"/> × <see cref="OverlapsPerCycle"/>.</summary>
    int DemuxBlockSize { get; }
}

/// <summary>
/// Implementation of <see cref="IPrecursorMaskCodec"/> that handles both overlapping DIA and
/// MSX experiments. Port of cpp's <c>PrecursorMaskCodec</c>.
/// </summary>
public sealed class PrecursorMaskCodec : IPrecursorMaskCodec
{
    /// <summary>Tunable parameters for cycle/overlap inference.</summary>
    public sealed class Params
    {
        /// <summary>Whether the data was acquired with variable fill times (the per-precursor
        /// mask is weighted by <c>MultiFillTime / 1000</c> when true).</summary>
        public bool VariableFill { get; init; }

        /// <summary>Tolerance (in m/z) used to decide whether two window boundaries are "the same".</summary>
        public double MinimumWindowSize { get; init; } = 0.2;

        /// <summary>Drop edge isolation segments not covered at the same multiplicity as the bulk
        /// of the cycle.</summary>
        public bool RemoveNonOverlappingEdges { get; init; }
    }

    private readonly Params _params;
    private readonly List<DemuxIsolationWindow> _isolationWindows = new();
    private readonly List<int> _edgeWindowsToRemove = new();
    private int _spectraPerCycle;
    private int _precursorsPerSpectrum;
    private int _overlapsPerSpectrum;

    /// <inheritdoc/>
    public int NumDemuxWindows => _isolationWindows.Count;

    /// <inheritdoc/>
    public IReadOnlyList<int> DemuxWindowEdgesRemoved => _edgeWindowsToRemove;

    /// <inheritdoc/>
    public int SpectraPerCycle => _spectraPerCycle;

    /// <inheritdoc/>
    public int PrecursorsPerSpectrum => _precursorsPerSpectrum;

    /// <inheritdoc/>
    public int OverlapsPerCycle => _overlapsPerSpectrum;

    /// <inheritdoc/>
    public int DemuxBlockSize => _spectraPerCycle * _precursorsPerSpectrum * _overlapsPerSpectrum;

    /// <summary>Constructs a codec by inferring the demux scheme from <paramref name="spectrumList"/>.</summary>
    public PrecursorMaskCodec(ISpectrumList spectrumList, Params? p = null)
    {
        ArgumentNullException.ThrowIfNull(spectrumList);
        _params = p ?? new Params();
        ReadDemuxScheme(spectrumList);
    }

    /// <inheritdoc/>
    public Vector<double> GetMask(Spectrum spectrum, double weight = 1.0)
    {
        var mask = DenseVector.Create(DemuxBlockSize, 0.0);
        FillMask((i, v) => mask[i] = v, spectrum, weight);
        return mask;
    }

    /// <inheritdoc/>
    public void GetMask(Spectrum spectrum, Matrix<double> m, int rowNum, double weight = 1.0)
    {
        for (int j = 0; j < m.ColumnCount; j++) m[rowNum, j] = 0;
        FillMask((i, v) => m[rowNum, i] = v, spectrum, weight);
    }

    private void FillMask(Action<int, double> setter, Spectrum spectrum, double weight)
    {
        var indices = new List<int>();
        SpectrumToIndices(spectrum, indices);

        if (_params.VariableFill)
        {
            // Per-index weighting from the MultiFillTime user param on each precursor.
            var indexedDemuxWindows = new List<DemuxWindow>(indices.Count);
            foreach (var idx in indices)
                indexedDemuxWindows.Add(_isolationWindows[idx].Window);

            foreach (var p in spectrum.Precursors)
            {
                var precursorWindow = new DemuxWindow(p);
                for (int i = 0; i < indices.Count; i++)
                {
                    if (precursorWindow.ContainsCenter(indexedDemuxWindows[i]))
                    {
                        double fillTime = p.UserParam("MultiFillTime").ValueAs<double>();
                        setter(indices[i], weight * fillTime / 1000.0);
                        break;
                    }
                }
            }
        }
        else
        {
            foreach (var idx in indices) setter(idx, weight);
        }
    }

    /// <inheritdoc/>
    public void SpectrumToIndices(Spectrum spectrum, List<int> indices)
    {
        if (spectrum.Precursors.Count != _precursorsPerSpectrum)
            throw new InvalidOperationException(
                "SpectrumToIndices() Number of precursors in this spectrum differs from the demultiplexing scheme.");

        indices.Clear();
        var overlappingWindows = new List<DemuxWindow>(spectrum.Precursors.Count);
        foreach (var precursor in spectrum.Precursors)
            overlappingWindows.Add(new DemuxWindow(precursor));
        overlappingWindows.Sort();

        int searchLowerBound = 0;
        foreach (var window in overlappingWindows)
        {
            for (int searchIdx = searchLowerBound; searchIdx < _isolationWindows.Count; searchIdx++)
            {
                var candidate = _isolationWindows[searchIdx].Window;
                if (window.MzHigh <= candidate.MzLow) break;
                if (window.ContainsCenter(candidate))
                {
                    indices.Add(searchIdx);
                    searchLowerBound = searchIdx + 1;
                }
            }
        }

        if (indices.Count == 0)
            throw new InvalidOperationException("SpectrumToIndices() found no demux windows for this spectrum.");

        if (indices.Count != _overlapsPerSpectrum * _precursorsPerSpectrum)
            throw new InvalidOperationException(
                "SpectrumToIndices() Number of demultiplexing windows changed. Minimum window size or " +
                "boundary tolerance may be set too low.");
    }

    /// <inheritdoc/>
    public DemuxIsolationWindow GetIsolationWindow(int i) => _isolationWindows[i];

    private void ReadDemuxScheme(ISpectrumList spectrumList)
    {
        IdentifyCycle(spectrumList, _isolationWindows);
        IdentifyOverlap(_isolationWindows);
    }

    private void IdentifyCycle(ISpectrumList spectrumList, List<DemuxIsolationWindow> demuxWindows)
    {
        const string Ms2MissingPrecursorError = "IdentifyCycle() MS2 spectrum is missing precursor information.";

        // First MS2 spectrum sets the expected precursors-per-spectrum count.
        int index = 0;
        Spectrum? firstMs2 = null;
        for (; index < spectrumList.Count; index++)
        {
            var spec = spectrumList.GetSpectrum(index, getBinaryData: false);
            if (spec.Params.CvParamValueOrDefault(CVID.MS_ms_level, 0) != 2) continue;
            if (spec.Precursors.Count == 0) throw new InvalidOperationException(Ms2MissingPrecursorError);
            _precursorsPerSpectrum = spec.Precursors.Count;
            firstMs2 = spec;
            break;
        }
        if (firstMs2 is null)
            throw new InvalidOperationException("IdentifyCycle() No MS2 scans found for this experiment.");

        // Walk forward until the cycle has been observed twice over (mappedAlready > 2 * |precursorMap|).
        // Map key is the rounded iso-center (cpp <c>prec_to_string</c>).
        var precursorMap = new Dictionary<string, Precursor>();
        int mappedAlready = 0;
        for (; index < spectrumList.Count && mappedAlready <= 2 * precursorMap.Count; index++)
        {
            var spec = spectrumList.GetSpectrum(index, getBinaryData: false);
            if (spec.Params.CvParamValueOrDefault(CVID.MS_ms_level, 0) != 2) continue;
            if (spec.Precursors.Count == 0) throw new InvalidOperationException(Ms2MissingPrecursorError);
            if (spec.Precursors.Count != _precursorsPerSpectrum)
                throw new InvalidOperationException(
                    "IdentifyCycle() Number of precursors is varying between MS2 scans. Cannot infer demultiplexing scheme.");

            foreach (var p in spec.Precursors)
            {
                string key = DemuxHelpers.PrecToString(p);
                if (!precursorMap.ContainsKey(key))
                {
                    mappedAlready = 0;
                    precursorMap[key] = p;
                }
                else
                {
                    mappedAlready++;
                }
            }
        }
        if (mappedAlready <= 2 * precursorMap.Count)
            throw new InvalidOperationException(
                "IdentifyCycle() Could not determine demultiplexing scheme. Too few spectra to determine the number of precursor windows.");

        // Sort keys numerically (cpp uses stof comparator).
        var sortedKeys = new List<string>(precursorMap.Keys);
        sortedKeys.Sort((a, b) => double.Parse(a, System.Globalization.CultureInfo.InvariantCulture)
            .CompareTo(double.Parse(b, System.Globalization.CultureInfo.InvariantCulture)));

        demuxWindows.Clear();
        demuxWindows.Capacity = sortedKeys.Count;
        foreach (var key in sortedKeys)
            demuxWindows.Add(DemuxIsolationWindow.FromPrecursor(precursorMap[key]));

        if (_precursorsPerSpectrum == 0)
            throw new InvalidOperationException("IdentifyCycle() Number of precursors per spectrum is 0.");
        _spectraPerCycle = demuxWindows.Count / _precursorsPerSpectrum;
        if (_spectraPerCycle == 0)
            throw new InvalidOperationException("IdentifyCycle() Number of spectra per cycle is 0.");
    }

    private void IdentifyOverlap(List<DemuxIsolationWindow> isolationWindows)
    {
        if (isolationWindows.Count <= 1) return;

        long minimumWindowSize = IsoWindowHasher.Hash(_params.MinimumWindowSize);

        // Collect every distinct boundary m/z (hashed) as a sorted set.
        var demuxBoundaries = new SortedSet<DemuxBoundary>();
        foreach (var w in isolationWindows)
        {
            demuxBoundaries.Add(new DemuxBoundary(w.LowMz));
            demuxBoundaries.Add(new DemuxBoundary(w.HighMz));
        }

        // Merge nearby boundaries within minimumWindowSize. Direct port of cpp's iterator-based
        // walk: each iteration consumes 1 boundary (far-apart case) or 2 boundaries (merge case).
        // Cpp's post-loop push uses the last value of lowMZ, NOT the last array element — they
        // can differ when a merge consumed the final pair.
        var exactBoundaries = new List<DemuxBoundary>();
        var boundaryArray = new DemuxBoundary[demuxBoundaries.Count];
        demuxBoundaries.CopyTo(boundaryArray);
        int low = 0, high = 1;
        while (high < boundaryArray.Length)
        {
            if (boundaryArray[high].MzHash - boundaryArray[low].MzHash > minimumWindowSize)
            {
                exactBoundaries.Add(boundaryArray[low]);
                low++; high++;
            }
            else
            {
                double avg = (boundaryArray[high].Mz + boundaryArray[low].Mz) / 2.0;
                exactBoundaries.Add(new DemuxBoundary(avg));
                low += 2; high += 2;
            }
        }
        // Cpp pushes lowMZ->mz after the loop. lowMZ may be one past the last consumed boundary
        // when the loop ended on a merge that ate the final pair — in that case there's nothing
        // left to add.
        if (low < boundaryArray.Length)
            exactBoundaries.Add(boundaryArray[low]);

        // Build the candidate sub-windows from adjacent boundary pairs.
        var possibleWindows = new List<DemuxIsolationWindow>(exactBoundaries.Count - 1);
        for (int i = 0; i + 1 < exactBoundaries.Count; i++)
            possibleWindows.Add(new DemuxIsolationWindow(exactBoundaries[i].Mz, exactBoundaries[i + 1].Mz));

        // For each precursor isolation window, mark which sub-windows it covers.
        // Use a list-of-equivalence-classes structure (cpp uses std::multiset<IsolationWindow>).
        var usedWindows = new List<DemuxIsolationWindow>();
        foreach (var iso in isolationWindows)
            foreach (var sub in possibleWindows)
                if (iso.Window.ContainsCenter(sub.Window))
                    usedWindows.Add(sub);
        usedWindows.Sort();

        // Walk groups of equal-window entries; the largest group size is the overlap multiplicity.
        var returnWindows = new List<DemuxIsolationWindow>();
        int maxCount = 0;
        int idx = 0;
        while (idx < usedWindows.Count)
        {
            var current = usedWindows[idx];
            int count = 1;
            while (idx + count < usedWindows.Count && usedWindows[idx + count].Window == current.Window) count++;
            if (count > maxCount) maxCount = count;
            if (_params.RemoveNonOverlappingEdges)
            {
                returnWindows.Add(current);
                if (count != maxCount) _edgeWindowsToRemove.Add(returnWindows.Count - 1);
            }
            else
            {
                returnWindows.Add(current);
            }
            idx += count;
        }

        if (_params.RemoveNonOverlappingEdges)
        {
            // Re-scan: now that maxCount is known, redo the edge marking with the final value.
            _edgeWindowsToRemove.Clear();
            int outIdx = 0;
            int srcIdx = 0;
            while (srcIdx < usedWindows.Count)
            {
                int count = 1;
                while (srcIdx + count < usedWindows.Count && usedWindows[srcIdx + count].Window == usedWindows[srcIdx].Window) count++;
                if (count != maxCount)
                {
                    System.Console.Error.WriteLine(
                        $"Dropping non-overlapping edge isolation window: {usedWindows[srcIdx].LowMz}, {usedWindows[srcIdx].HighMz}");
                    _edgeWindowsToRemove.Add(outIdx);
                }
                outIdx++;
                srcIdx += count;
            }
        }

        _overlapsPerSpectrum = maxCount;
        if (_overlapsPerSpectrum == 0)
            throw new InvalidOperationException("IdentifyOverlap() Number of demux windows is 0.");

        isolationWindows.Clear();
        isolationWindows.AddRange(returnWindows);
    }

    /// <summary>Helper struct for splitting / merging isolation-window edges during overlap
    /// analysis. Cpp's <c>PrecursorMaskCodec::DemuxBoundary</c>.</summary>
    private readonly record struct DemuxBoundary(double Mz, long MzHash) : IComparable<DemuxBoundary>
    {
        public DemuxBoundary(double mz) : this(mz, IsoWindowHasher.Hash(mz)) { }
        public int CompareTo(DemuxBoundary other) => MzHash.CompareTo(other.MzHash);
    }
}
